protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnMapperTest.java

```diff
@@ -0,0 +1,305 @@
+            testee.getUid(0).isPresent()
+        ).isFalse();
+    }
+
+    @Test
+    public void getUidShouldTheCorrespondingUidIfItExist() {
+        testee.addUid(messageUid1);
+
+        Optional<MessageUid> result = testee.getUid(1);
+
+        assertThat(result.isPresent()).isTrue();
+        assertThat(result.get()).isEqualTo(messageUid1);
+    }
+
+    @Test
+    public void getFirstUidShouldReturnEmptyIfNoMessage() {
+        assertThat(testee.getFirstUid().isPresent()).isFalse();
+    }
+
+    @Test
+    public void getLastUidShouldReturnEmptyIfNoMessage() {
+        assertThat(testee.getLastUid().isPresent()).isFalse();
+    }
+
+    @Test
+    public void getFirstUidShouldReturnFirstUidIfAtLeastOneMessage() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+
+        Optional<MessageUid> result = testee.getFirstUid();
+
+        assertThat(result.isPresent()).isTrue();
+        assertThat(result.get()).isEqualTo(messageUid1);
+    }
+
+    @Test
+    public void getLastUidShouldReturnLastUidIfAtLeastOneMessage() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+
+        Optional<MessageUid> result = testee.getLastUid();
+
+        assertThat(result.isPresent()).isTrue();
+        assertThat(result.get()).isEqualTo(messageUid2);
+    }
+
+    @Test
+    public void getMsnShouldReturnAbsentIfNoCorrespondingMessage() {
+        testee.addUid(messageUid1);
+
+        assertThat(testee.getMsn(messageUid2).isPresent()).isFalse();
+    }
+
+    @Test
+    public void getMsnShouldReturnMessageNumberIfUidIsThere() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+
+        Optional<Integer> result = testee.getMsn(messageUid2);
+        assertThat(result.isPresent()).isTrue();
+        assertThat(result.get()).isEqualTo(2);
+    }
+
+    @Test
+    public void getNumMessageShouldReturnZeroIfNoMapping() {
+        assertThat(testee.getNumMessage()).isEqualTo(0);
+    }
+
+    @Test
+    public void getNumMessageShouldReturnTheNumOfMessage() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+
+        assertThat(testee.getNumMessage()).isEqualTo(2);
+    }
+
+    @Test
+    public void isEmptyShouldReturnTrueIfNoMapping() {
+        assertThat(testee.isEmpty()).isTrue();
+    }
+
+    @Test
+    public void isEmptyShouldReturnFalseIfNoMapping() {
+        testee.addUid(messageUid1);
+
+        assertThat(testee.isEmpty()).isFalse();
+    }
+
+    @Test
+    public void clearShouldClearMapping() {
+        testee.addUid(messageUid1);
+
+        testee.clear();
+
+        assertThat(testee.isEmpty()).isTrue();
+    }
+
+    @Test
+    public void addUidShouldKeepMessageNumberContiguous() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid2);
+        testee.addUid(messageUid3);
+        testee.addUid(messageUid4);
+        testee.addUid(messageUid2);
+
+        assertThat(messageNumberAreContiguous(testee)).isTrue();
```

Comment:
Because I need to read an other method (that is not even tested) to know what you assert